src/System.Net.Security/src/System/Net/_SslSessionsCache.cs

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+
+namespace System.Net.Security
+{
+    // Implements SSL session caching mechanism based on a static table of SSL credentials.
+    internal static class SslSessionsCache
+    {
+        private const int c_CheckExpiredModulo = 32;
+        private static Hashtable s_CachedCreds = new Hashtable(32);
+
+        //
+        // Uses cryptographically strong certificate thumbprint comparison.
+        //
+        private struct SslCredKey
+        {
+            private byte[] _CertThumbPrint;
+            private int _AllowedProtocols;
+            private EncryptionPolicy _EncryptionPolicy;
+            private int _HashCode;
+
+            //
+            // SECURITY: X509Certificate.GetCertHash() is virtual hence before going here,
+            //           the caller of this ctor has to ensure that a user cert object was inspected and
+            //           optionally cloned.
+            //
+            internal SslCredKey(byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
+            {
+                _CertThumbPrint = thumbPrint == null ? Array.Empty<byte>() : thumbPrint;
+                _HashCode = 0;
+                if (thumbPrint != null)
+                {
+                    _HashCode ^= _CertThumbPrint[0];
+                    if (1 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[1] << 8);
+                    if (2 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[2] << 16);
+                    if (3 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[3] << 24);
+                }
+                _HashCode ^= (int)allowedProtocols;
+                _HashCode ^= (int)encryptionPolicy;
+                _AllowedProtocols = allowedProtocols;
+                _EncryptionPolicy = encryptionPolicy;
+            }
+
+            public override int GetHashCode()
+            {
+                return _HashCode;
+            }
+
+            public static bool operator ==(SslCredKey sslCredKey1,
+                                            SslCredKey sslCredKey2)
+            {
+                if ((object)sslCredKey1 == (object)sslCredKey2)
+                {
+                    return true;
+                }
+                if ((object)sslCredKey1 == null || (object)sslCredKey2 == null)
```

Comment:
SslCredKey is a struct.  Seems like the first if will end up boxing the keys, such that they won't have reference equality, and the second if will box them and then never find them null.  So allocation that ends up being a nop?